server/src/main/java/org/elasticsearch/indices/analysis/wrappers/StableApiWrappers.java

```diff
@@ -204,10 +206,47 @@ private static <T> T createInstance(
         Environment environment
     ) {
         try {
-            Constructor<T> constructor = clazz.getConstructor();
-            return constructor.newInstance();
-        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
-            throw new IllegalStateException("cannot create instance of " + clazz, e);
+
+            Constructor<?>[] constructors = clazz.getConstructors();
+            if (constructors.length > 1) {
+                throw new IllegalStateException("Plugin can only have one public constructor.");
+            }
+            Constructor<?> constructor = constructors[0];
+            if (constructor.getParameterCount() == 0) {
+                return (T) constructor.newInstance();
+            } else {
+                InjectSettings injectSettings = constructor.getAnnotation(InjectSettings.class);
+                if (injectSettings != null) {
+                    Class<?>[] parameterTypes = constructor.getParameterTypes();
+                    Object[] parameters = new Object[parameterTypes.length];
+                    for (int i = 0; i < parameterTypes.length; i++) {
+                        Object settings = createSettings(parameterTypes[i], indexSettings, nodeSettings, analysisSettings, environment);
+                        parameters[i] = settings;
+                    }
+                    return (T) constructor.newInstance(parameters);
+                } else {
+                    throw new IllegalStateException("Missing @Inject annotation for constructor with settings.");
+                }
+            }
+
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalStateException("Cannot create instance of " + clazz, e);
```

Comment:
should this be IllegalArgumentException?
the exception will be thrown upon the settings and plugin component creation. So when calling indices.analyze or creating a mapping